### PR TITLE
tests: Add bootc end-to-end validation test

### PR DESCRIPTION
**Pull request by martinpitt**
See https://issues.redhat.com/browse/RHEL-88396 and https://github.com/linux-system-roles/tox-lsr/pull/202. As mentioned in the latter, I also have branches for less trivial roles, firewall (passing) and postgresql (spotted a bug). But I'd like to get this reviewed first, as iterating is cheap.

After we agree to the structure, I'll submit the general structure (first four commits) to .github.git, and send PRs for firewall and postgresql.

 - [x] Land https://github.com/linux-system-roles/tox-lsr/pull/202 and drop the XXX commit
 - [ ] Discuss and land this or #59

## Summary by Sourcery

Add end-to-end bootc validation testing and update CI to support bootc-image scenarios

New Features:
- Add bootc end-to-end validation test suite for QEMU container images

CI:
- Install tox-lsr from the qemu-bootc branch
- Pin Podman to 5.x on Ubuntu to maintain compatibility with the bootc-image-builder
- Adjust tox-lsr invocation to skip and later execute bootc-e2e tests separately
- Add a CI step to run bootc-e2e tests in QEMU for container-bootc images

Tests:
- Introduce tests/tests_bootc_e2e.yml Ansible playbook defining the bootc end-to-end validation test

## Summary by Sourcery

Add bootc end-to-end validation testing and enhance CI workflows to install updated dependencies, pin Podman, and run bootc-e2e tests separately in QEMU

New Features:
- Add a bootc end-to-end validation test suite for QEMU container-bootc images

CI:
- Bump tox-lsr dependency to v3.10.0 across GitHub Actions workflows
- Pin Podman to version 5.x on Ubuntu for compatibility with the bootc-image-builder
- Skip bootc-e2e tests in regular runs and introduce a dedicated CI step to execute them in QEMU

Tests:
- Introduce tests/tests_bootc_e2e.yml playbook for end-to-end validation of bootc container images

### CODE REVIEW

File: .github/workflows/qemu-kvm-integration-tests.yml [hunk 1]
```diff
@@ -117,7 +139,7 @@ jobs:
         if: steps.check_platform.outputs.supported && startsWith(matrix.scenario.env, 'qemu')
         run: >-
           tox -e ${{ matrix.scenario.env }} -- --image-name ${{ matrix.scenario.image }} --make-batch
-          --log-level debug --skip-tags tests::infiniband,tests::nvme,tests::scsi
+          --log-level debug $TOX_ARGS --skip-tags tests::bootc-e2e
```
[the remark this answers is not in the document]
Comment (richm) [reply .github/workflows/qemu-kvm-integration-tests.yml:142]: We should also use `TOX_ARGS: "--skip-tags tests::infiniband,tests::nvme,tests::scsi"` for this.

File: .github/workflows/qemu-kvm-integration-tests.yml [hunk 2]
```diff
@@ -164,6 +186,32 @@ jobs:
...
+        if: steps.check_platform.outputs.supported &&
+            startsWith(matrix.scenario.env, 'container') &&
+            endsWith(matrix.scenario.image, '-bootc')
+        run: |
+          set -euxo pipefail
+          env=$(echo "${{ matrix.scenario.env }}" | sed 's/^container-/qemu-/')
+
+          for t in $(grep -wl tests::bootc-e2e tests/tests_*.yml); do
+              test=$(basename ${t%.yml})
+              image_file="$(pwd)/tests/tmp/$test/qcow2/disk.qcow2"
```
Comment (richm) [inline .github/workflows/qemu-kvm-integration-tests.yml:199]: Ah - this is where the `/tmp/` directory is used.  Hmm - not sure how to get the `$LSR_TOX_ENV_TMP_DIR` in this context - this may also be problematic when testing the collections format, because the test directory is `tests/$rolename/` and not `....$rolename/tests/` there

File: .github/workflows/qemu-kvm-integration-tests.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinpitt) [reply .github/workflows/qemu-kvm-integration-tests.yml:142]: Because that's the exact value of the defined `TOX_ARGS`. The workflow defines that at the top but never actually used it. And it seems sensible to be able to parameterize it. Otherwise we could drop the unused .definition at the top and repeat the value (or not, it's not super important for bootc e2e tests)

File: tests/tests_bootc_e2e.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        - name: Run the role
+          include_role:
+            name: linux-system-roles.sudo
+          vars:
+            sudo_check_if_configured: false
+
+        - name: Create QEMU deployment
+          delegate_to: localhost
+          command: "{{ lsr_scriptdir }}/bootc-buildah-qcow.sh {{ ansible_host }}"
+          changed_when: true
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply tests/tests_bootc_e2e.yml:21]: No, I don't think so. In the three roles I've looked at (sudo, postgresql, firewall) , I found that there's enough reasons to keep the bootc e2e tests separate. They have a different structure, and are *very* expensive, so running all tests in that mode will just take too long. IMHO it's better to pick a moderately complex representative case to make sure that services etc. start up, and leave the finer details/corner cases to the usual qemu/bootc build-only tests. See https://issues.redhat.com/browse/RHEL-88396 for details.

File: .github/workflows/qemu-kvm-integration-tests.yml (hunk 2 above)
Comment (martinpitt) [reply .github/workflows/qemu-kvm-integration-tests.yml:199]: This isn't `$LSR_TOX_ENV_TMP_DIR` at the moment, but `OUTPUT="./tmp/$BUILDAH_ID"` in bootc-buildah-qcow.sh (in the tox-lsr PR). We can of course change it on both sides.

**Comment by martinpitt** [comment]
Meh, now [F41 fails with](https://github.com/linux-system-roles/sudo/actions/runs/15386084109/job/43284901718?pr=58) "Installing to filesystem: Creating ostree deployment: invalid reference format". This is reproducible locally. Reported as https://github.com/osbuild/bootc-image-builder/issues/949 , and I don't know a workaround. Let's give this a few days, in the meantime this can get reviewed. If it lasts longer, we can skip the e2e test on F41.

**Comment by martinpitt** [comment]
Updated to tox 3.10.0. Most probably we'll land #59 instead, but then this at least serves as a worked example how to do a new test.

**Review by richm — APPROVED** [review]

**Comment by martinpitt** [comment]
Landed #59.
